File: Program1/Java/src/FixedPointListController.java
import java.util.Scanner;

public class FixedPointListController {
    FixedPointList fixedPointList;
    public FixedPointListController(){
        fixedPointList = new FixedPointList();


    }
    public static void main(String[] args) {
        FixedPointListController controller = new FixedPointListController();
        controller.run();
    }

    public void run(){
        String response;
        Scanner scanner = new Scanner(System.in);
        do {
            response = execute(scanner.nextLine());
            System.out.print(response + '\n');

        }
        while (!response.equals("Normal termination of program1."));
    }


    /**
     * Handle raw line from keyboard. Will remove carriage return before passing
     * to other handlers.
     *
     * @param line - Raw input from keyboard. -> scanner.readLine()
     * @return Response from system to display to user
     * */
    public String execute(String line){
        line = line.replace("\n", "");
        if(line.contains(" "))
        {
            int indexOfSpace = line.indexOf(" ");
            String args = line.substring(indexOfSpace + 1);
            String commandOnly = line.substring(0,indexOfSpace);

            return this.handleCommand(commandOnly, args);
        }
        else
        {
            return  this.handleCommand(line);
        }
    }


    private String handleCommand(String command){
        switch (command.toLowerCase()) {
            case "a" -> {
                return "Arg missing. You must specify the number to delete from the list";
            }
            case "q" -> {
                return "Arg missing. You must specify the new q value";
            }
            case "d" -> {
                return "Arg missing. You must specify the number to append to the list";
            }
            case "p" -> {
                return "All fixed-point numbers in the list are:" + fixedPointList.toString();

            }
            case "s" -> {
                return "The sum is " + fixedPointList.sumAll().toString() + ".";
            }
            case "x" -> {
                return "Normal termination of program1.";
            }
            default -> {
                return command + " is not a valid command!";
            }
        }


    }

    private String handleCommand(String command, String arg){

        switch (command.toLowerCase()) {
            case "a" -> {

                double dArg = Double.parseDouble(arg);
                FixedPointNumber num = new FixedPointNumber(dArg, fixedPointList.getCurrentQValue());
                fixedPointList.add(num);
                return num.toString() + " was added to the list.";
            }
            case "d" -> {

                double dArg = Double.parseDouble(arg);
                FixedPointNumber num = new FixedPointNumber(dArg, fixedPointList.getCurrentQValue());
                boolean result = fixedPointList.delete(num);

                if (result) {
                    return num.toString() + " was deleted from the list.";
                } else {
                    return "No value equal to " + num.toString() + " in the list.";
                }
            }
            case "q" -> {
                int newQ = Integer.parseInt(arg);
                fixedPointList.setCurrentQValue(newQ);
                return "Current q_value was changed to " + fixedPointList.getCurrentQValue() + ".";
            }
            case "p", "s", "x" -> {
                return handleCommand(command);
            }
            default -> {
                return command + " is not a valid command!";
            }
        }
    }
}

File: Program1/Ruby/src/FixedPointListController.rb
# frozen_string_literal: true

require './FixedPointList'

class FixedPointListController
  attr_accessor :num_list

  def initialize
    super
    @num_list = FixedPointList.new
  end

  def run
    continue = true
    while continue

      command = gets
      response = execute(command)
      puts response
      continue = response != 'Normal termination of program1.'

    end
  end

  def execute(command)
    has_arg = command.match(/\s/)

    if has_arg
      split_command = command.split(' ')
      command = split_command[0]
      response = handle_command(command, split_command[1])
    else
      response = handle_command(command)
    end

    response
  end

  def handle_command(command, arg = nil)
    command_lower = command.downcase
    if command_lower == 'a'
      handle_add(arg)
    elsif command_lower == 'q'
      handle_q_change(arg)
    elsif command_lower == 'd'
      handle_delete(arg)
    elsif command_lower == 'p'
      entries = @num_list.to_s.to_s
      return "All fixed-point numbers in the list are:\n#{entries}" if entries != ''

      'All fixed-point numbers in the list are:'
    elsif command_lower == 's'
      "The sum is #{@num_list.sum}."
    elsif command_lower == 'x'
      'Normal termination of program1.'
    else
      "#{command} is not a valid command!"
    end
  end

  def handle_add(arg)
    if !arg.nil?
      new_fixed_point = FixedPointNumber.new(arg.to_f, @num_list.current_q)
      @num_list.add(new_fixed_point)
      "#{new_fixed_point} was added to the list."
    else
      'Arg missing. You must specify the number to append to the list'
    end
  end

  def handle_q_change(arg)
    if !arg.nil?
      @num_list.current_q = arg.to_i
      "Current q_value was changed to #{@num_list.current_q}."
    else
      'Arg missing. You must specify the new q value'
    end
  end

  def handle_delete(arg)
    if !arg.nil?
      new_fixed_point = FixedPointNumber.new(arg.to_f, @num_list.current_q)
      result = @num_list.delete(new_fixed_point)
      if result
        "#{new_fixed_point} was deleted from the list."
      else
        "No value equal to #{new_fixed_point} in the list."
      end
    else
      'Arg missing. You must specify the number to append to the list'
    end
  end

  private :handle_command, :handle_add, :handle_delete, :handle_q_change
end
